cpp/src/arrow/json/parser.cc

```diff
@@ -0,0 +1,982 @@
+          field_builder.nullable = f->nullable();
+          Cast<Kind::kObject>(*builder)->AddField(f->name(), field_builder);
+        }
+        return Status::OK();
+      }
+      default:
+        return Status::NotImplemented("invalid builder type");
+    }
+  }
+
+  Status AppendNull() {
+    if (!builder_.nullable) {
+      return ParseError("a required field was null");
+    }
+    switch (builder_.kind) {
+      case Kind::kNull: {
+        // increment null count stored inline
+        // update the parent, since changing builder_ doesn't affect parent
+        auto parent = builder_stack_.back();
+        if (parent.kind == Kind::kArray) {
+          auto list_builder = Cast<Kind::kArray>(parent);
+          DCHECK_EQ(list_builder->value_builder(), builder_);
+          builder_.index += 1;
+          list_builder->value_builder(builder_);
+        } else {
+          auto struct_builder = Cast<Kind::kObject>(parent);
+          DCHECK_EQ(struct_builder->field_builder(field_index_), builder_);
+          builder_.index += 1;
+          struct_builder->field_builder(field_index_, builder_);
+        }
+        return Status::OK();
+      }
+      case Kind::kBoolean:
+        return Cast<Kind::kBoolean>(builder_)->AppendNull();
+      case Kind::kNumber:
+        return Cast<Kind::kNumber>(builder_)->AppendNull();
+      case Kind::kString:
+        return Cast<Kind::kString>(builder_)->AppendNull();
+      case Kind::kArray:
+        return Cast<Kind::kArray>(builder_)->AppendNull();
+      case Kind::kObject: {
+        auto root = builder_;
+        auto struct_builder = Cast<Kind::kObject>(builder_);
+        RETURN_NOT_OK(struct_builder->AppendNull());
+        for (int i = 0; i != struct_builder->num_fields(); ++i) {
+          builder_ = struct_builder->field_builder(i);
+          RETURN_NOT_OK(AppendNull());
+        }
+        builder_ = root;
+        return Status::OK();
+      }
+      default:
+        return Status::NotImplemented("invalid builder Kind");
+    }
+  }
+
+  Status AppendBool(bool value) {
+    constexpr auto kind = Kind::kBoolean;
+    if (builder_.kind != kind) {
+      return IllegallyChangedTo(kind);
+    }
+    return Cast<kind>(builder_)->Append(value);
+  }
+
+  template <Kind::type kind>
+  Status AppendScalar(string_view scalar) {
+    if (builder_.kind != kind) {
+      return IllegallyChangedTo(kind);
+    }
+    auto index = static_cast<int32_t>(scalar_values_builder_.length());
+    RETURN_NOT_OK(Cast<kind>(builder_)->Append(index));
+    return scalar_values_builder_.Append(scalar);
+  }
+
+  Status StartObjectImpl() {
+    constexpr auto kind = Kind::kObject;
+    if (builder_.kind != kind) {
+      return IllegallyChangedTo(kind);
+    }
+    auto struct_builder = Cast<kind>(builder_);
+    absent_fields_stack_.Push(struct_builder->num_fields(), true);
+    PushStacks();
+    return struct_builder->Append();
+  }
+
+  bool SetFieldBuilder(string_view key) {
```

Comment:
There's a number of functions in here that could use comments to explain what they do, or where they are used. If you look at where this one is used it's obvious, like "we've just seen a JSON key, get the builder ready for the value"